tests/lib/rules/no-reference-type-as-default-prop.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * @fileoverview Prevent usage of referential-type variables as default param in functional component
+ * @author Chang Yan
+ */
+
+'use strict';
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const RuleTester = require('eslint').RuleTester;
+const rule = require('../../../lib/rules/no-reference-type-as-default-prop');
+
+const parserOptions = {
+  ecmaVersion: 2018,
+  sourceType: 'module',
+  ecmaFeatures: {
+    jsx: true
+  }
+};
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({parserOptions});
+const MESSAGE_ID = 'forbiddenTypeDefaultParam';
+
+ruleTester.run('no-reference-type-as-default-prop', rule, {
+  valid: [
+    `
+      function Foo({
+        bar = emptyFunction,
+      }) {}
+  `,
+    `
+    function Foo({
+      bar = emptyFunction,
+      ...rest
+    }) {}
+  `,
+    `
+    function Foo({
+      bar = 1,
+      baz = 'hello',
+    }) {}
+  `,
+    `
+    function Foo(props) {}
+  `,
+    `
+    function Foo(props) {}
+
+    Foo.defaultProps = {
+      bar: () => {}
+    }
+  `,
+    `
+    const Foo = () => {};
+  `,
+    `
+    const Foo = ({bar = 1}) => {};
+  `
+  ],
+  invalid: [
```

Comment:
let's make sure each of these test cases is duplicated for the default parser, babel-eslint, and the typescript parsers.